pyiron_base/generic/flattenedstorage.py

```diff
@@ -469,41 +467,76 @@ def to_hdf(self, hdf, group_name="flat_storage"):
+                write_array(k, a, hdf_arrays)
+
+            hdf_arrays = hdf_s_lst.open("chunk_arrays")
+            for k, a in self._per_chunk_arrays.items():
+                write_array(k, a, hdf_arrays)
 
     def from_hdf(self, hdf, group_name="flat_storage"):
         with hdf.open(group_name) as hdf_s_lst:
             version = hdf_s_lst.get("HDF_VERSION", "0.0.0")
-            num_chunks = hdf_s_lst["num_chunks"] or hdf_s_lst["num_structures"]
-            num_elements = hdf_s_lst["num_elements"] or hdf_s_lst["num_atoms"]
+            try:
+                num_chunks = hdf_s_lst["num_chunks"]
+                num_elements = hdf_s_lst["num_elements"]
+            except KeyError:
+                num_chunks = hdf_s_lst["num_structures"]
+                num_elements = hdf_s_lst["num_atoms"]
+
             self._num_chunks_alloc = self.num_chunks = self.current_chunk_index = num_chunks
             self._num_elements_alloc = self.num_elements = self.current_element_index = num_elements
 
-            with hdf_s_lst.open("arrays") as hdf_arrays:
-                for k in hdf_arrays.list_nodes():
-                    a = np.array(hdf_arrays[k])
-                    if a.dtype.char == "S":
-                        # if saved as bytes, we wrote this as an encoded unicode string, so manually decode here
-                        # TODO: string arrays with shape != () not handled
-                        a = np.array([s.decode("utf8") for s in a],
-                                    # itemsize of original a is four bytes per character, so divide by four to get
-                                    # length of the orignal stored unicode string; np.dtype('U1').itemsize is just a
-                                    # platform agnostic way of knowing how wide a unicode charater is for numpy
-                                    dtype=f"U{a.dtype.itemsize//np.dtype('U1').itemsize}")
-                    if a.shape[0] == self._num_elements_alloc:
-                        self._per_element_arrays[k] = a
-                    elif a.shape[0] == self._num_chunks_alloc:
-                        self._per_chunk_arrays[k] = a
+            def read_array(name, hdf):
+                a = np.array(hdf[name])
+                if a.dtype.char == "S":
+                    # if saved as bytes, we wrote this as an encoded unicode string, so manually decode here
+                    # TODO: string arrays with shape != () not handled
+                    a = np.array([s.decode("utf8") for s in a],
+                                # itemsize of original a is four bytes per character, so divide by four to get
+                                # length of the orignal stored unicode string; np.dtype('U1').itemsize is just a
+                                # platform agnostic way of knowing how wide a unicode charater is for numpy
+                                dtype=f"U{a.dtype.itemsize//np.dtype('U1').itemsize}")
```

Comment:
I didn't test anything yet, but why do you define read_array and write_array inside other functions? Does

Reply:
I nested the functions, because I don't want them as methods, since they a) don't access the instance and b) would bloat the class API.  Alternatively I could make them private top-level functions, but then the definition is far away from the usage and it's more difficult to read.  Creating nested functions is [not](https://stackoverflow.com/questions/14055843/what-is-the-performance-overhead-of-nested-functions) for [free](https://stackoverflow.com/questions/7839632/is-there-an-overhead-when-nesting-functions-in-python/7839697#7839697), but it's also not expensive.  If it ever comes up in a profile, we can still pull it out.

Reply:
I am not against defining the function in a nested way, however, I would prefer to have function definitions at the beginning of the main function and not being nested again in the `with open... as f` context.